### PR TITLE
Implement logloss interface

### PR DESCRIPTION
**Pull request by avolkov-intel**
-Add objective function algorithm in oneapi/dal/algo
-Use existing logloss primitive to compute logloss using GPU
-Integrate DAAL logloss implementation to compute logloss using CPU

### CODE REVIEW

File: cpp/oneapi/dal/algo/logloss_objective/common.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace oneapi::dal::logloss_objective::detail {
+
+namespace v1 {
+
+template <typename Task>
+class descriptor_impl : public base {
+public:
+    explicit descriptor_impl(double l1_regularization_coefficient = 0,
+                             double l2_regularization_coefficient = 0)
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/logloss_objective/common.cpp:28]:
```suggestion
    explicit descriptor_impl(double l1_regularization_coefficient = 0.0,
                             double l2_regularization_coefficient = 0.0)
```

File: cpp/oneapi/dal/algo/logloss_objective/common.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            : l1_regularization_coefficient(l1_regularization_coefficient),
+              l2_regularization_coefficient(l2_regularization_coefficient) {}
+    double l1_regularization_coefficient = 0;
+    double l2_regularization_coefficient = 0;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/logloss_objective/common.cpp:32]:
```suggestion
    double l1_regularization_coefficient = 0.0;
    double l2_regularization_coefficient = 0.0;
```

File: cpp/oneapi/dal/algo/objective_function/backend/cpu/compute_kernel.cpp [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    if (result.get_result_options().test(result_options::gradient)) {
+        arr_grad = array<Float>::empty(p + 1);
+        daal_gradient = interop::convert_to_daal_homogen_table(arr_grad, p + 1, 1);
+        daal_parameter.resultsToCompute |= daal_obj_fun::gradient;
+    }
+
+    if (result.get_result_options().test(result_options::hessian)) {
+        arr_hess = array<Float>::empty((p + 1) * (p + 1));
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/backend/cpu/compute_kernel.cpp:86]: Please check integer overflow by multiplication

File: cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+}
+
+template <typename Float>
+result_t compute_kernel_dense_batch_impl<Float>::operator()(
+    const detail::descriptor_base<task_t>& desc,
+    const input_t& input) {
+    ONEDAL_ASSERT(input.get_data().has_data());
+    ONEDAL_ASSERT(input.get_parameters().has_data());
+    ONEDAL_ASSERT(input.get_responses().has_data());
+    const auto data = input.get_data();
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp:53]:
```suggestion
    const auto* const data = input.get_data();
```

File: cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+    const auto params_nd = pr::table2ndarray_1d<Float>(q_, params, alloc::device);
+    const auto* const params_ptr = params_nd.get_data();
+
+    const auto responses_nd_big = pr::table2ndarray_1d<std::int32_t>(q_, responses, alloc::device);
+
+    auto result = compute_result<task_t>{}.set_result_options(desc.get_result_options());
+
+    auto [out, out_e] = pr::ndarray<Float, 1>::zeros(q_, { p + 2 }, sycl::usm::alloc::device);
+    out_e.wait_and_throw();
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp:78]: Too many calls of `.wait_and_throw()`. Please review it and pass events correctly

File: cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        out_hessian = pr::ndarray<Float, 2>::empty(q_, { p + 1, p + 1 }, sycl::usm::alloc::device);
+        ans_hessian = pr::ndarray<Float, 2>::empty(q_, { p + 1, p + 1 }, sycl::usm::alloc::device);
+        auto fill_event = fill(q_, ans_hessian, Float(0), {});
+        fill_event.wait_and_throw();
+    }
+
+    auto probabilities_big = pr::ndarray<Float, 1>::empty(q_, { bsz }, sycl::usm::alloc::device);
+
+    for (std::int64_t i = 0; i < n; i += bsz) {
+        std::int64_t cursize = std::min(n - i, bsz);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp:104]:
```suggestion
        const std::int64_t cursize = std::min(n - i, bsz);
```

File: cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+    if (desc.get_result_options().test(result_options::hessian)) {
+        out_hessian = pr::ndarray<Float, 2>::empty(q_, { p + 1, p + 1 }, sycl::usm::alloc::device);
+        ans_hessian = pr::ndarray<Float, 2>::empty(q_, { p + 1, p + 1 }, sycl::usm::alloc::device);
+        auto fill_event = fill(q_, ans_hessian, Float(0), {});
+        fill_event.wait_and_throw();
+    }
+
+    auto probabilities_big = pr::ndarray<Float, 1>::empty(q_, { bsz }, sycl::usm::alloc::device);
+
+    for (std::int64_t i = 0; i < n; i += bsz) {
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp:103]: What about using `uniform_blocking`?

File: cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+                                                  data_nd,
+                                                  responses_nd,
+                                                  probabilities,
+                                                  out_loss,
+                                                  Float(0),
+                                                  Float(0),
+                                                  { prob_e, fill_event });
+                loss_event.wait_and_throw();
+
+                const auto* const out_ptr = out_loss.get_data();
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp:156]: Const `out` pointer? Are you sure?

Comment (KulikovNikita) [reply cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp:156]: Please review it

File: cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+                }
+            });
+        });
+        update_event.wait_and_throw();
+    }
+    if (desc.get_result_options().test(result_options::gradient)) {
+        auto update_event = q_.submit([&](sycl::handler& cgh) {
+            const auto range = oneapi::dal::backend::make_range_1d(p + 1);
+            cgh.parallel_for(range, [=](sycl::id<1> idx) {
+                ans_grad_ptr[idx[0]] /= n;
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp:232]: `n` value is common for all work items. Please move it out

1. Create a variable outside of the kernel. Let say `const Float inv_n = Float(1) / n;`
2. Use it inside of the kernel

File: cpp/oneapi/dal/algo/objective_function/common.cpp [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+/*
+result_option_id get_packed_gradient_id() {
+    return result_option_id{ result_option_id::make_by_index(3) };
+}
+
+result_option_id get_packed_hessian_id() {
+    return result_option_id{ result_option_id::make_by_index(4) };
+}
+*/
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/common.cpp:43]: Please mark it as `TODO`

File: cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+                                      out_loss,
+                                      Float(0),
+                                      Float(0),
+                                      { fill_event });
+    const auto* const out_ptr = out_loss.get_data();
+    auto* const ans_loss_ptr = ans_loss.get_mutable_data();
+    return q_.submit([&](sycl::handler& cgh) {
+        cgh.depends_on({ loss_event, prev_iter });
+        cgh.single_task([=] {
+            ans_loss_ptr[0] += out_ptr[0];
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/backend/gpu/compute_kernel_dense_batch_impl_dpc.cpp:180]:
```suggestion
            *ans_loss_ptr += *out_ptr;
```

File: cpp/oneapi/dal/algo/objective_function/detail/compute_ops.hpp [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+                result.get_value().get_column_count() != 1) {
+                throw domain_error(msg::incorrect_output_table_size());
+            }
+        }
+
+        if (result.get_result_options().test(result_options::gradient)) {
+            if (!result.get_gradient().has_data()) {
+                throw domain_error(msg::gradient_is_not_provided());
+            }
+            if (result.get_gradient().get_row_count() != p + 1 ||
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/detail/compute_ops.hpp:86]:
```suggestion
            if (result.get_gradient().get_row_count() != (p + 1) ||
```

File: cpp/oneapi/dal/algo/objective_function/detail/compute_ops.hpp [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+                throw domain_error(msg::incorrect_output_table_size());
+            }
+        }
+
+        if (result.get_result_options().test(result_options::hessian)) {
+            if (!result.get_hessian().has_data()) {
+                throw domain_error(msg::hessian_is_not_provided());
+            }
+            if (result.get_hessian().get_row_count() != p + 1 ||
+                result.get_hessian().get_column_count() != p + 1) {
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/detail/compute_ops.hpp:97]:
```suggestion
            if (result.get_hessian().get_row_count() != (p + 1) ||
                result.get_hessian().get_column_count() != (p + 1)) {
```

File: cpp/oneapi/dal/algo/objective_function/test/fixture.hpp [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+
+namespace oneapi::dal::objective_function::test {
+
+namespace pr = dal::backend::primitives;
+namespace te = dal::test::engine;
+namespace de = dal::detail;
+namespace obj_fun = oneapi::dal::objective_function;
+namespace lg = oneapi::dal::logloss_objective;
+
+std::mt19937 rnd(2007);
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/algo/objective_function/test/fixture.hpp:35]: Please, do not use global variables. Even in tests

**Comment by avolkov-intel** [comment]
https://ecmd.jf.intel.com/commander/link/jobDetails/jobs/edc41359-8e15-f1ac-afdb-a4bf010d0e2e

**Comment by avolkov-intel** [comment]
@samir-nasibli could you review this PR, please?

File: examples/oneapi/cpp/source/objective_function/logloss_dense_batch.cpp [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+                                        result_options::hessian);
+
+    auto result = dal::compute(desc, data, params, labels);
+
+    std::cout << "Logistic loss\n" << result.get_value() << std::endl;
+
+    std::cout << "Gradient\n" << result.get_gradient() << std::endl;
+
+    std::cout << "Hessian\n" << result.get_hessian() << std::endl;
+}
```
Comment (Alexsandruss) [inline examples/oneapi/cpp/source/objective_function/logloss_dense_batch.cpp:53]: Add newline at end of file

**Review by KulikovNikita — APPROVED** [review]

**Pull request merged**